source/opt/dead_branch_elim_pass.cpp

```diff
@@ -73,260 +72,215 @@ bool DeadBranchElimPass::GetConstInteger(uint32_t selId, uint32_t* selVal) {
-    uint32_t condId;
-    if (!GetSelectionBranch(*bi, &br, &mergeInst, &condId)) continue;
+  auto get_parent_block = [this](uint32_t id) {
+    return context()->get_instr_block(get_def_use_mgr()->GetDef(id));
+  };
 
-    // If constant condition/selector, replace conditional branch/switch
-    // with unconditional branch and delete merge
-    uint32_t liveLabId;
-    if (br->opcode() == SpvOpBranchConditional) {
+  // Mark live blocks reachable from the entry. Simplify constant branches and
+  // switches as you proceed to limit the number of live blocks. Be careful not
+  // to eliminate backedges even if they are dead, but the header is live.
+  std::unordered_map<ir::BasicBlock*, ir::BasicBlock*> continues;
+  bool modified = false;
+  std::unordered_set<ir::BasicBlock*> liveBlocks;
+  std::vector<ir::BasicBlock*> stack;
+  stack.push_back(&*func->begin());
+  while (!stack.empty()) {
+    ir::BasicBlock* block = stack.back();
+    stack.pop_back();
+
+    // Live blocks doubles as visited set.
+    if (!liveBlocks.insert(block).second) continue;
+
+    uint32_t contId = block->ContinueBlockIdIfAny();
+    if (contId != 0) continues[get_parent_block(contId)] = block;
+
+    ir::Instruction* terminator = &*block->tail();
+    uint32_t liveLabId = 0;
+    // Check if the terminator has a single valid successor.
+    if (terminator->opcode() == SpvOpBranchConditional) {
       bool condVal;
-      if (!GetConstCondition(condId, &condVal)) continue;
-      liveLabId = (condVal == true)
-                      ? br->GetSingleWordInOperand(kBranchCondTrueLabIdInIdx)
-                      : br->GetSingleWordInOperand(kBranchCondFalseLabIdInIdx);
-    } else {
-      assert(br->opcode() == SpvOpSwitch);
-      // Search switch operands for selector value, set liveLabId to
-      // corresponding label, use default if not found
+      if (GetConstCondition(terminator->GetSingleWordInOperand(0u), &condVal)) {
+        liveLabId = terminator->GetSingleWordInOperand(
+            condVal ? kBranchCondTrueLabIdInIdx : kBranchCondFalseLabIdInIdx);
+      }
+    } else if (terminator->opcode() == SpvOpSwitch) {
       uint32_t selVal;
-      if (!GetConstInteger(condId, &selVal)) continue;
-      uint32_t icnt = 0;
-      uint32_t caseVal;
-      br->ForEachInOperand(
-          [&icnt, &caseVal, &selVal, &liveLabId](const uint32_t* idp) {
-            if (icnt == 1) {
-              // Start with default label
-              liveLabId = *idp;
-            } else if (icnt > 1) {
-              if (icnt % 2 == 0) {
-                caseVal = *idp;
-              } else {
-                if (caseVal == selVal) liveLabId = *idp;
-              }
+      if (GetConstInteger(terminator->GetSingleWordInOperand(0u), &selVal)) {
+        // Search switch operands for selector value, set liveLabId to
+        // corresponding label, use default if not found.
+        uint32_t icnt = 0;
+        uint32_t caseVal;
+        terminator->ForEachInOperand([
```

Comment:
I know this was code that was already there, but, since we are doing a large change, could we change this to a for-loop.  Plus it can avoid all of the logic with `icnt`, which I always find hard to follow.